docs/general/adapters.md

```diff
@@ -194,6 +194,43 @@ The user could pass in `include=**`.
 
 We recommend filtering any user-supplied includes appropriately.
 
+#### Include Parameters
+
+You may wish to include related resources based on the `include` request
+parameter (See [fetching_includes](http://jsonapi.org/format/#fetching-includes). You can opt-in to this pattern:
+
+```ruby
+class PostSerializer < ActiveModel::Serializer
+  associations_via_include_param(true)
```

Comment:
So, the problem this solves is wanting to make a relationship link without loading the relationship, or to render the relationship id and type without loading the relationship?  Somewhere else we had a longish discussion about doing it like JR does 

(I like to distinguish between the problem the code is solving and the implementation itself.  I think we can do this with less 'shotgun surgery ')

B mobile phone

> On Jun 9, 2016, at 9:18 PM, Lee Richmond <notifications@github.com> wrote:
> 
> In docs/general/adapters.md:
> 
> > @@ -194,6 +194,43 @@ The user could pass in `include=**`.
> >  
> >  We recommend filtering any user-supplied includes appropriately.
> >  
> > +#### Include Parameters
> > +
> > +You may wish to include related resources based on the `include` request
> > +parameter (See [fetching_includes](http://jsonapi.org/format/#fetching-includes). You can opt-in to this pattern:
> > +
> > +```ruby
> > +class PostSerializer < ActiveModel::Serializer
> > +  associations_via_include_param(true)
> Good questions! I absolutely agree the naming gets confusing here and glad to take any suggestions.
> 
> What does _associations_via_include_param connote?
> 
> That you're opting in to this feature. The biggest reason why you might not want to opt-in is if you want the relationships payload to always include id/type regardless of the ?include parameter. The biggest reason you want to opt-in is to avoid the performance penalty fetching the id/types will incur. IMO this is the 80%+ use case and is the current behavior of json-api-resources
> 
> What's the difference between include_directive and current_include_directive?
> 
> include_directive is the current code - this is the IncludeDirective passed to render. It looks something like { posts: { comments: { author: {} }, tags: {} }, tags: {} }. current_include_directive is the slice of that hash being processed by the current serializer (so CommentSerializer would see { author: {} }. We need to differentiate between these two because the same serializer might be used multiple times at a different nesting level (like tags in this example).
> 
> What's the difference between current_include_directive and full_include_directive
> 
> Good question. full_include_directive is the same as what I've noted above as include_directive - the current behavior of the gem, the full directive passed to render. I could make it full_include_directive everywhere if that would make it more readable.
> 
> Why are there two include directives now?
> 
> Explained above.
> 
> Why is there a param include_directive introduced that defaults to {} when no arg passed?
> 
> This is the current code actually...it's just providing a default for when no include arg is passed to render. And this default can be customized, I think so people can default to including all relationships (*).
> 
> What does load_data connote?
> 
> We're in the relationships block here, where this gem provides functionality like link - "here's the proc to fire when fetching links data". Currently the return value of the block can be used to customize the value of the relationship:
> 
> class PostSerializer < ActiveModel::Serializer
>   has_many :comments do
>     link :related do
>       href "/posts/#{object.id}/relationships/comments"
>     end
> 
>     Comment.not_deleted
>   end
> end
> The problem is, this PR wants to process that link proc without firing that DB query that's going to get all undeleted comments. So we wrap that data fetching part in a proc, this way we can run the block but delay execution of data fetching. Open to any alternate names here, I've also suggested just forcing the user to return a proc.
> 
> More discussion on why we need to delay the execution is in the comments above.
> 
> What's the difference between include_data and load_data?
> 
> include_data is current code - it means, "should we include the data part of the payload for this relationship?". Currently users set this to false/true. load_data is explained above - the proc to fetch the relationship.
> 
> Why is the consequence of include_data? a choice between @_load_data.call and include_data_for?
> 
> If the relationship data key should be in the response (include_date?), we need to load this relationship. So either use a custom proc (inline-relationship functionality of AMS), or just go through the normal codepath.
> 
> Hope this helps, @bf4, appreciate your time looking at this!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>